Restructure client requirements (#291)
MikeBishop/dns-alt-svc: draft-ietf-dnsop-svcb-https.md

# Client behavior {#client-behavior}

"SVCB resolution" is the process of enumerating the priority-ordered endpoints
for a service, as performed by the client.  It MUST be implemented as follows:

1. Let $QNAME be the service name plus appropriate prefixes for the
   scheme (see {{svcb-names}}).

2. Issue a SVCB query for $QNAME.

3. If an AliasMode SVCB record is returned, set $QNAME to its TargetName (without
   additional prefixes) and loop back to step 2,
   subject to chain length limits and loop detection heuristics (see
   {{client-failures}}).

4. If one or more "compatible" ({{mandatory}}) ServiceMode records are returned,
   these represent the alternative endpoints.

5. Otherwise, SVCB resolution has failed.

This procedure does not rely on any recursive or authoritative DNS server to
comply with this specification or have any awareness of SVCB.

Once SVCB resolution is completed, the client proceeds with connection
establishment.  Clients SHOULD try higher-priority alternatives first, with

> fallback to lower-priority alternatives.  Clients MUST issue AAAA and/or A

Reviewer: ```suggestion
fallback to lower-priority alternatives.  Clients issue AAAA and/or A
```

queries for the selected TargetName, and MAY choose between them using an
approach such as Happy Eyeballs {{!HappyEyeballsV2=RFC8305}}.

A client is called "SVCB-reliant" if it connects only to the indicated
ServiceMode endpoints, and "SVCB-optional" otherwise.  Clients for pre-existing
protocols SHALL implement SVCB-optional behavior (except as noted in

> {{client-failures}} and {{ech-client-behavior}}).

Reviewer: This definition of SVCB-reliant doesn't match with my understanding.  I think that what you want to do is note that there are two reasons to make an A/AAAA query in parallel.  The primary one being that the client might be willing to fall back to A/AAAA queries if there are no SVCB records, but you might also want to allow for the possibility that a TargetName form SVCB record references the same name and so there might be an advantage in making the A/AAAA queries in parallel to allow for that outcome.

(Unfortunately, if you *need* SVCB, you need the TargetName form, and you don't get to predict when that will appear, so you basically have to make the parallel query always once you decide that it is worth the parallel.)

You mention the latter below, which is good, but I think that it applies here.

Author: I'm not sure I understand your concern here, but I hope these changes have clarified it somewhat.

Reviewer: I think that the key difference between reliant and optional is whether the information in the ServiceMode is critical to operation of the client.

Related question: does an SVCB-optional client switch to SVCB-reliant if it encounters an AliasMode record?  Or, can it follow to the target name and just consume A/AAAA (or whatever) instead, without ever having seen a ServiceMode record?

Author: > I think that the key difference between reliant and optional is whether the information in the ServiceMode is critical to operation of the client.

Agreed.  How does that differ from the definition given here?

> Related question: does an SVCB-optional client switch to SVCB-reliant if it encounters an AliasMode record?

No.

> Or, can it follow to the target name and just consume A/AAAA (or whatever) instead, without ever having seen a ServiceMode record?

Yes.  That's important, to support use of AliasMode pointing to a hosting provider that doesn't have a SVCB record.

Additionally, as currently specified, SVCB-optional clients are generally encouraged (and in some cases required) to fall back to non-SVCB connection if all the SVCB alternative endpoints are tried and they all fail. (See #256 for some debate about that.)

Reviewer: >  How does that differ from the definition given here?

The definition here concentrates on what the client does (connect to the ServiceForm endpoint), rather than the reasons for the same.

Does the draft explain the AliasForm-CNAME isomorphism?  There's a lot of text, but I can't see anything there.  Seems important.

Author: > The definition here concentrates on what the client does (connect to the ServiceForm endpoint), rather than the reasons for the same.

OK, I've reversed the polarity of the definitions.  Does that help?

> Does the draft explain the AliasForm-CNAME isomorphism?

It's not really an isomorphism.  [The AliasMode section](https://tools.ietf.org/html/draft-ietf-dnsop-svcb-https-02#section-2.4.2) discusses the relationship in some detail.  I'm happy to improve the description but I think it's out of scope for this PR.

SVCB-optional clients SHOULD issue in parallel any other DNS queries that might
be needed for connection establishment.  SVCB-optional clients SHALL append an
alternative endpoint consisting of the final value of $QNAME, the authority

> endpoint's port number, and no SvcParams, and MAY append the authority endpoint.

Reviewer: I don't understand this.  Are you trying to back-fill a default value for SVCB in the case you don't need one?  Why not just say "if the endpoint does not require SVCB, then it can use A/AAAA records when SVCB is not available"?

Author: There are actually _two_ "default" fallbacks listed here, in a specific order and with different normativity.  The main motivation is to support using AliasMode to point to a hosting provider that doesn't use SVCB.

I've adjusted this text to improve clarity (and avoid assuming that non-SVCB connection means A/AAAA, since it could also be SRV or something else).

Reviewer: As the text still says that you issue A/AAAA when you hit a ServiceMode record, I'm now more confused by this.

I'm still totally short on clarity here.  What is being appended to?  And I still don't know what "the authority endpoint" is or what it is appended to.  Are you saying that if you exhaust SVCB lookups and you are don't need the ServiceForm information, you could just pretend to have received a SVCB ServiceForm record that consists of $QNAME, the port, and no SvcParams?  Why say that at all?  What advantage does this have over letting the client fall back to non-SVCB resolution?

Author: > As the text still says that you issue A/AAAA when you hit a ServiceMode record, I'm now more confused by this.

Connection using a ServiceMode record always requires an address lookup for TargetName.  Connection _not_ using SVCB doesn't strictly require an address lookup ... who knows what the non-SVCB connection procedure is for an arbitrary protocol?

> What is being appended to?

The priority-ordered list of alternative endpoints.

> And I still don't know what "the authority endpoint" is or what it is appended to.

That's defined in [the definitions section](https://tools.ietf.org/html/draft-ietf-dnsop-svcb-https-02#section-1.4).

> Are you saying that if you exhaust SVCB lookups and you are don't need the ServiceForm information, you could just pretend to have received a SVCB ServiceForm record that consists of $QNAME, the port, and no SvcParams?

Yes.

> Why say that at all?

Mostly to handle the case where there is an AliasMode record but no ServiceMode record.  Adding this endpoint unconditionally for fallback is not necessary but seems like it could improve reliability (one more endpoint to try) without expanding the DNS response.

Reviewer 2: Whereas with a precise read of the text does cover the "the case where there is an AliasMode record but no ServiceMode record" it is not obvious to a reader as shown by both Martin and I not quite getting this until reading a few times.  At a minimum we should have a parenthetical which concretely calls out this case.

Some important optimizations are discussed in {{optimizations}}
to avoid additional latency in comparison to ordinary AAAA/A lookups.

## Handling resolution failures {#client-failures}

If SVCB resolution fails due to a SERVFAIL error, transport error, or timeout,
and DNS exchanges between the client and the recursive resolver are
cryptographically protected (e.g. using TLS {{!DoT=RFC7858}} or HTTPS
{{!DoH=RFC8484}}), a SVCB-optional client SHOULD terminate the connection like

> a SVCB-reliant client would.  Otherwise, an active attacker

Reviewer: terminate the connection or terminate the connection attempt?

Author: Connection attempt.  Reworded.

   {{client-failures}}).

4. If one or more "compatible" ({{mandatory}}) ServiceMode records are returned,
   these represent the alternative endpoints.

5. Otherwise, SVCB resolution has failed.

This procedure does not rely on any recursive or authoritative DNS server to
comply with this specification or have any awareness of SVCB.

Once SVCB resolution is completed, the client proceeds with connection
establishment.  Clients SHOULD try higher-priority alternatives first, with
fallback to lower-priority alternatives.  Clients issue AAAA and/or A
queries for the selected TargetName, and MAY choose between them using an
approach such as Happy Eyeballs {{!HappyEyeballsV2=RFC8305}}.

A client is called "SVCB-optional" if it can connect without the use of
ServiceMode records, and "SVCB-reliant" otherwise.  Clients for pre-existing
protocols SHALL implement SVCB-optional behavior (except as noted in
{{client-failures}} and {{ech-client-behavior}}).

SVCB-optional clients SHOULD issue in parallel any other DNS queries that might
be needed for connection establishment.  SVCB-optional clients SHALL append an
alternative endpoint consisting of the final value of $QNAME, the authority

> endpoint's port number, and no SvcParams, to be attempted before falling back
> to non-SVCB connection modes.

Reviewer: ```suggestion
endpoint's port number, and no SvcParams, to the list of alternative endpoints, which is 
attempted before falling back to non-SVCB connection modes.
```

Does this need to be explicitly ServiceMode?

# Client behavior {#client-behavior}

A SVCB-aware client selects an endpoint for a service using the following
procedure:
"SVCB resolution" is the process of enumerating the priority-ordered endpoints
for a service, as performed by the client.  It MUST be implemented as follows:

1. Let $ADDR_QNAME be the service name.  Let $SVCB_QNAME be the service name
   plus appropriate prefixes for the scheme (see {{svcb-names}}).
1. Let $QNAME be the service name plus appropriate prefixes for the
   scheme (see {{svcb-names}}).

2. In parallel, issue AAAA/A queries for $ADDR_QNAME
   and a SVCB query for $SVCB_QNAME.

>    The answers for these may or may not include CNAME pointers

Reviewer 2: Do we want to keep a reference to potential of CNAMEs in-place, or this is adequately covered elsewhere and in examples?

Author: OK, added to Step 3.

5. Otherwise, SVCB resolution has failed, and the list of known endpoints is
   empty.

This procedure does not rely on any recursive or authoritative DNS server to
comply with this specification or have any awareness of SVCB.

Once SVCB resolution has concluded, the client proceeds with connection
establishment.  Clients SHOULD try higher-priority alternatives first, with
fallback to lower-priority alternatives.  Clients issue AAAA and/or A
queries for the selected TargetName, and MAY choose between them using an
approach such as Happy Eyeballs {{!HappyEyeballsV2=RFC8305}}.

A client is called "SVCB-optional" if it can connect without the use of
ServiceMode records, and "SVCB-reliant" otherwise.  Clients for pre-existing
protocols SHALL implement SVCB-optional behavior (except as noted in
{{client-failures}} and {{ech-client-behavior}}).

SVCB-optional clients SHOULD issue in parallel any other DNS queries that might
be needed for connection establishment.  SVCB-optional clients SHALL append an
alternative endpoint consisting of the final value of $QNAME, the authority
endpoint's port number, and no SvcParams, to the list of alternative endpoints, which is 

> attempted before falling back to non-SVCB connection modes.

Reviewer 2: The case somewhat ambiguous here is the scenario where an AliasMode record points to a name with only A/AAAA records.  It may help to call this out explicitly.  Which may be clarifying that each time $QNAME changes, SVCB-optional clients SHOULD issue AAAA/A queries.  We may also want some text such as:  "Note that SVCB-optional clients should be prepared to handle the case where the TargetName of an AliasMode SVCB RR has only AAAA/A records and no ServiceMode record."

Author: OK, I've added a clarification for that case.

# Client behavior {#client-behavior}

"SVCB resolution" is the process of enumerating the priority-ordered endpoints

> for a service, as performed by the client.  It MUST be implemented as follows:

Reviewer: ```suggestion
for a service, as performed by the client.  SVCB resolution is implemented as follows:
```

I think that the MUST here is too strong.  The W3C and WHATWG, who specialize in writing algorithmic specifications, only ever say that the outcome should be equivalent to (or indistinguishable from) the algorithm as described.

With these optimizations in place, and conforming DNS servers,
using SVCB does not add network latency to connection setup.

To improve performance when using a non-conforming recursive resolver,
clients SHOULD issue speculative A and AAAA queries based on a predicted value
of TargetName.  In the absence of better information, clients SHOULD predict
that TargetName will be the service name for the initial query and "." for

> subsequent queries (see {zone-performance}).

Reviewer: This isn't something you recommend.  You just need to say that the reason you issue A/AAAA in parallel is that you are predicting that TargetName is present and that TargetName is ".", which is what {{zone-performance}} (two curlies!) recommends for those that provision records.

Author: {{zone-performance}} actually didn't recommend that.  I've moved this recommendation from here to there.

If none of the SVCB records are consistent
with any active or in-progress connection,

> clients must proceed with connection establishment as described in

Reviewer: ```suggestion
clients proceed with connection establishment as described in
```

If Alt-Svc connection fails, these clients SHOULD fall back to the HTTPS RR
client connection procedure ({{client-behavior}}).

Clients that implement support for ECH MUST perform the HTTPS RR query first,
and MUST only make use of Alt-Svc when operating in SVCB-optional mode (see
{{ech-client-behavior}}).

This specification does not alter the DNS queries performed when connecting

> to an Alt-Svc hostname (typically A and/or AAAA only).

Reviewer: Citation?

Author: Clarified

Reviewer: I was noting that this might need citation, though the clarification is fine.

## Client behavior {#ech-client-behavior}

The SVCB-optional client behavior specified in {{client-behavior}} permits clients
to fall back to a direct connection if all SVCB options fail.  This behavior is
not suitable for ECH, because fallback would negate the privacy benefits of
ECH.  Accordingly, ECH-capable SVCB-optional clients MUST implement the following
behavior for connection establishment:

1. Perform SVCB resolution as described in {{client-behavior}}.
2. If SVCB resolution succeeded, and all alternative endpoints have an "echconfig"

>    key, use SVCB-reliant connection establishment.

Reviewer 2: Does this mean that ECH-capable SVCB-optional clients can't use Alt-Svc until we define an Alt-Svc-bis that addresses relevant issues?

Author: No.  ECH-capable SVCB-optional clients can use Alt-Svc unless all returned ServiceForm records include `echconfig`.

Reviewer 2: Is there a reason we're missing what had been the previous steps 2+3?  (ie, to handle the case where the conditional in step 2 doesn't hold?)

Author: It's effectively moved to line 1054 (in the Alt-Svc section, instead of talking about Alt-Svc in the ECH section).

Reviewer 2: In-general this numbered list is still confusing since it doesn't have an "else" clause to 2.  Perhaps switching it to be a bulletted list or just some flat text would be better?

Reviewer 2: Perhaps:

Accordingly, ECH-capable SVCB-optional clients MUST use SVCB-reliant connection establishment if SVCB resolution succeeded (following {{client-behavior}}), and if all alternative endpoints have an "echconfig" key.

Author: OK, done.
